runtime/lua/vim/ui.lua

```diff
@@ -167,29 +167,65 @@ function M.open(path)
   return vim.system(cmd, opts), nil
 end
 
---- Gets the URL at cursor, if any.
-function M._get_url()
-  if vim.bo.filetype == 'markdown' then
-    local range = vim.api.nvim_win_get_cursor(0)
-    vim.treesitter.get_parser():parse(range)
-    -- marking the node as `markdown_inline` is required. Setting it to `markdown` does not
-    -- work.
-    local current_node = vim.treesitter.get_node { lang = 'markdown_inline' }
-    while current_node do
-      local type = current_node:type()
-      if type == 'inline_link' or type == 'image' then
-        local child = assert(current_node:named_child(1))
-        return vim.treesitter.get_node_text(child, 0)
+--- Returns all URLs at cursor, if any.
+--- @return string[]
+function M._get_urls()
+  local urls = {}
+
+  local bufnr = vim.api.nvim_get_current_buf()
+  local cursor = vim.api.nvim_win_get_cursor(0)
+  local row = cursor[1] - 1
+  local col = cursor[2]
+  local extmarks = vim.api.nvim_buf_get_extmarks(bufnr, -1, { row, col }, { row, col }, {
+    details = true,
+    type = 'highlight',
+    overlap = true,
+  })
+  for _, v in ipairs(extmarks) do
+    local details = v[4]
+    if details.url then
+      urls[#urls + 1] = details.url
+    end
+  end
+
+  local highlighter = vim.treesitter.highlighter.active[bufnr]
+  if highlighter then
+    local range = { row, col, row, col }
+    local ltree = highlighter.tree:language_for_range(range)
+    local lang = ltree:lang()
+    local query = vim.treesitter.query.get(lang, 'highlights')
+    if query then
+      local tree = ltree:tree_for_range(range)
+      for pattern, match, metadata in
+        query:iter_matches(tree:root(), bufnr, row, row + 1, { all = true })
```

Comment:
If https://github.com/neovim/neovim/pull/30193 merges first, be sure to remove `{ all = true }`